pkg/labels/consul_aggregator.go

```diff
@@ -0,0 +1,174 @@
+package labels
+
+import (
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/kp/consulutil"
+	"github.com/square/p2/pkg/logging"
+
+	"github.com/square/p2/Godeps/_workspace/src/github.com/Sirupsen/logrus"
+	"github.com/square/p2/Godeps/_workspace/src/github.com/hashicorp/consul/api"
+	"github.com/square/p2/Godeps/_workspace/src/k8s.io/kubernetes/pkg/labels"
+)
+
+// minimum required time between retrievals of label subtrees.
+var AggregationRateCap = 10 * time.Second
+
+// linked list of watches with control channels
+type selectorWatch struct {
+	selector labels.Selector
+	resultCh chan *[]Labeled
+	canceled chan struct{}
+	next     *selectorWatch
+}
+
+func (s *selectorWatch) append(w *selectorWatch) {
+	if w == nil {
+		return
+	}
+	if s.next != nil {
+		s.next.append(w)
```

Comment:
Here and in `delete()`, do you know if the Go compiler will apply any tail call optimization to limit the size of the stack while traversing data structures?